test/WebSites/Basic/Program.cs

```diff
@@ -1,20 +1,42 @@
 ﻿using Microsoft.AspNetCore.Hosting;
+
+#if NETCOREAPP3_0
+using Microsoft.Extensions.Hosting;
+#else
 using System.IO;
+#endif
 
 namespace Basic
 {
     public class Program
     {
         public static void Main(string[] args)
         {
-            var host = new WebHostBuilder()
+            CreateHostBuilderAndRun(args);
+        }
+
+        public static void CreateHostBuilderAndRun(string[] args) =>
+#if NETCOREAPP3_0
+            NC3_CreateHostBuilder(args).Build().Run();
+#else
+            NC2_CreateHostBuilder().Build().Run();
+
+#endif
+
+#if NETCOREAPP3_0
+        public static IHostBuilder NC3_CreateHostBuilder(string[] args) =>
+            Host.CreateDefaultBuilder(args)
+                .ConfigureWebHostDefaults(webBuilder =>
+                {
+                    webBuilder.UseStartup<Startup>();
+                });
+#else
+        public static IWebHostBuilder NC2_CreateHostBuilder() =>
+            new WebHostBuilder()
                 .UseKestrel()
                 .UseContentRoot(Directory.GetCurrentDirectory())
                 .UseIISIntegration()
-                .UseStartup<Startup>()
-                .Build();
-
-            host.Run();
-        }
+                .UseStartup<Startup>();
+#endif
     }
 }
```

Comment:
Same as above

Reply:
I'm halfway through the review and just don't feel comfortable with the proliferation of cross-targeting conditionals throughout the code. As I stated before, my preference is to target the lowest dependency versions that work in all cases, something which David Fowler seems to indicate is still possible. From https://github.com/aspnet/AspNetCore/issues/14393#issuecomment-535111340:
> Not really no. You don't need to target netcoreapp3.0 unless you're trying to use new features in netcoreapp3.0 (or if you end up hitting a breaking change between versions). The library author guidance is usually to target the lowest version that satisfies where you need to run. You only end up needing to multi-target when you want to take advantage of new functionality in the newer version or if you run into something truly incompatible between the 2 versions.


> Not really no. You don't need to target netcoreapp3.0 unless you're trying to use new features in netcoreapp3.0 (or if you end up hitting a breaking change between versions). The library author guidance is usually to target the lowest version that satisfies where you need to run. You only end up needing to multi-target when you want to take advantage of new functionality in the newer version or if you run into something truly incompatible between the 2 versions.

Reply:
If a project in the Swashbuckle needs to use new features in netcoreapp3.0 or is hitting breaking changes, as was the case in the projects that _already_ multi-targeted prior to this PR, then obviously multi-targeting is the only way. However, if this isn't the case, then I think I'd prefere to target netstandard2.1 and therefore minimize the need for cross-targeting conditionals

Reply:
TBH, `Swashbuclke` would really need/love to use both `UseRouting` and `UseEndpoints()` as all routes metadata are now properly handled by `UseRouting` and you would be able to use it latter on, and the middleware could then be `.MapSwagger()` in the endpoints.

I trully wonder if all the new `Endpoint Routing` stack could in fact help you, as you only want to support 1 code base, the difference between 2.x and 3.x are huge regarding the routing.

Starting January 2020, only the `2.1` will be "supported" (not 2.2 anymore), and `3.x`. The version `3.1` will be LTS.
The idea of this cross targeting will let you be able to still support LTS (`netstandard2.0`) and currently `3.0` that totally change the routing, this is why i attempts in this PR to leverage both side of the `Mvc` that works in their own way.
I do intend to take a look after this PR at :
* Swashbuckle for `UseEndpoints()`
* `System.Texty.Json` (default ?) vs external package for the `NewtonsoftApiModelResolver` (not sure how yet, but as of today `Newtonsoft` hard dependency in swagger might not let user move completly out ^^)
* Use `Static Content Asset` (new in 3.0) to serve front/static file from libs

Reply:
Understood that SB _could_ benefit from the newer features but rather than pre-empting the need for them, can we add those dependencies (and therefore cross-target conditionals) as necessary - i.e. in future PRs?

For the sample apps, I think I'd prefer for them to just target netcoreapp3.0. There purpose is to demonstrate different configuration options with Swashbuckle - not test coverage and not documenting upgrades. Those could be reasonable well covered by just having one sample app called `NetCore2`, which would ensure everything is backwards compatible and demo's setup for the older framework version

Reply:
Regarding your bullet points above - I'm already in the process of incorporating `System.Text.Json` so please don't start a PR on that as we'll be duplicating work 😉. My plan for this to support the new Json API by default. As part of this, I'll move `NewtonsoftApiModelResolver` to a separate package that's explicitly opt-in - basically mirroring the .NET Core 3.0 defaults.

Reply:
> Regarding your bullet points above - I'm already in the process of incorporating System.Text.Json so please don't start a PR 

Yup saw that this week end don't worry
also each bullet point would be separated PR but these can land latter, whereas this one intend to fully ensure that all dependencies will resolves `2.x` or `3.x` but not a mix of `2.x` + `3.x` for `aspnetcore3`.

If you're confident with `v5.0.0` this PR can land latter, i just wanted to give an idea of what i thought woul dhave been the best for `AspNetCore 3` stack.

How do you want to proceed ?

Reply:
So, I think these changes _should_ make their way into the Swsahbuckle code, I just wonder if it should happened on a more incremental/as-needed basis. If nothing else, smaller PR's that each solve a very specific problem will help me understand them better.

You clearly have a deep understanding of the .NET Core infrastructure and so I'm excited to have someone with that level of knowledge help transition the project so that it's optimized for the latest stable stack.

With that said, I think we could proceed with an incremental list of changes. For example:

1. Out-of-the-box support for `System.Text.Json` (in work)
2. Upgrade unit tests AND sample apps to use the latest and greatest 3.0 configuration, and retain an additional `NetCore2` sample app to ensure and demonstrate backwards compatibility
3. Leverage the new Endpoints API
4. Leverate static content assets for the UI projects

I expect 3 & 4 will involve some level of mult-targeting to continue supporting 2.0 and 3.0 stacks, and I think these would be the forcing functions to introduce those dependencies as neccessary.

As I'm currently working no 1 above, you could tackle any one (or all) of the others. Thoughts? (and thanks for all your input so far).

Reply:
@tebeco ^^^

Reply:
(finishing to fix all the test (less than 6), then will read and rewind everything properly.

I'll be able to have a stable branch to start from all test green ;)

Reply:
the 1., if i understand correctly, you are currently working on it
the 3. could only be done after the current PR to be fair as there will be lots of noise
the 4. is totally unrelated to 3. but would require this one

and to be honnest the current PR look like the 2.
as the cross targeting prove the double full compatibility of both stack with their respective dependencies
i just need to update the sample from 2.0 to 2.1
and attempt to use the `UseRouting` is the samples (not in swagger just for integration test so it prouve the double full compat)

Reply:
is that what you had in mind ?

Reply:
@domaindrivendev 

https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/1308#issuecomment-542552055